lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -83,19 +98,201 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
     }
   }
 
-  SILModule &M = OrigF->getModule();
-  Module *SM = M.getSwiftModule();
+  // Check if we have substitutions which replace generic type parameters with
+  // concrete types or unbound generic types.
+  bool HasConcreteGenericParams = false;
+  bool HasUnboundGenericParams = false;
+  for (auto &entry : InterfaceSubs.getMap()) {
+    // Check only the substitutions for the generic parameters.
+    // Ignore any dependent types, etc.
+    if (!isa<GenericTypeParamType>(entry.first))
+      continue;
+    if (entry.second->getCanonicalType()->hasArchetype()) {
+      HasUnboundGenericParams = true;
+      continue;
+    }
+    HasConcreteGenericParams = true;
+  }
+
+  if (!HasConcreteGenericParams) {
+    // All substititions are unbound.
+    DEBUG(llvm::dbgs() <<
+          "    Cannot specialize with all unbound interface substitutions.\n");
+    DEBUG(for (auto Sub : ParamSubs) {
+            Sub.dump();
+          });
+    return;
+  }
 
-  SubstitutedType = SILType::substFuncType(M, SM, InterfaceSubs.getMap(),
-                                           OrigF->getLoweredFunctionType(),
-                                           /*dropGenerics = */ true);
+  // We need a generic environment for the partial specialization.
+  if (HasUnboundGenericParams && !OrigF->getGenericEnvironment())
+    return;
+
+  // Substitutions to be used to create a new function type for the
+  // specialized function. If the the original substitution list mapped
+  // a given generic parameter to a concrete type, then do the same.
+  // Map all other generic parameters to themselves.
+  SubstitutionMap CallerInterfaceSubs;
+  if (!HasUnboundGenericParams)
+    CallerInterfaceSubs = InterfaceSubs;
+
+  if (HasUnboundGenericParams) {
+    auto OrigGenSig = OrigF->getLoweredFunctionType()->getGenericSignature();
+    auto OrigGenericEnv = OrigF->getGenericEnvironment();
+    SpecializedGenSig = OrigGenSig;
+
+    // Form a new generic signature based on the old one.
+    ArchetypeBuilder Builder(*SM, M.getASTContext().Diags);
+
+    // First, add the old generic signature.
+    Builder.addGenericSignature(OrigGenSig, nullptr);
+
+    // Checks if a given generic parameter type or a dependent type
+    // is mapped to a non-concrete type, i.e. a type that has
+    // archetypes. For dependent types, it also checks that its
+    // base generic parameter type is also mapped to a non-concrete type.
+    auto IsNonConcreteReplacementType =
+        [&InterfaceSubs](CanType Ty, Type DefaultReplacementTy) -> bool {
+      if (DefaultReplacementTy->getCanonicalType()->hasArchetype())
+        return true;
+      auto BaseTy = Ty;
+      while (auto DepTy = dyn_cast<DependentMemberType>(BaseTy)) {
+        BaseTy = DepTy->getBase()->getCanonicalType();
+      };
+      assert(isa<GenericTypeParamType>(BaseTy));
+      return InterfaceSubs.getMap()
+          .lookup(BaseTy.getPointer())
+          ->getCanonicalType()
+          ->hasArchetype();
```

Comment:
@slavapestov I added a comment about this check. I may need your help to understand why the code crashes compiling stdlib without this check.